presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,104 @@
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import io.airlift.log.Logger;
+import org.joda.time.DateTime;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static org.joda.time.DateTime.now;
+import static org.joda.time.Seconds.secondsBetween;
+
+@ThreadSafe
+public class ClusterMemoryLeakDetector
+{
+    private static final Logger log = Logger.get(ClusterMemoryLeakDetector.class);
+
+    // It may take some time to remove a query's memory reservations from the worker nodes, that's why
+    // we check to see whether some time has passed after the query finishes to claim that it is leaked.
+    private static final int DEFAULT_LEAK_CLAIM_DELTA_SEC = 60;
+
+    @GuardedBy("this")
+    private Set<QueryId> leakedQueries;
+
+    /**
+     * @param queryExecutionsSupplier All queries that the coordinator knows about.
+     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
+     */
+    void checkForMemoryLeaks(Supplier<List<QueryInfo>> queryInfoSupplier, Map<QueryId, Long> queryMemoryReservations)
+    {
+        requireNonNull(queryInfoSupplier);
+        requireNonNull(queryMemoryReservations);
+
+        ImmutableMap<QueryId, QueryInfo> queryIdToInfo = Maps.uniqueIndex(queryInfoSupplier.get(), QueryInfo::getQueryId);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> isLeaked(queryIdToInfo, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        synchronized (this) {
+            leakedQueries = ImmutableSet.copyOf(leakedQueryReservations.keySet());
+        }
+    }
+
+    private static boolean isLeaked(ImmutableMap<QueryId, QueryInfo> queryIdToInfo, QueryId queryId)
+    {
+        QueryInfo queryInfo = queryIdToInfo.get(queryId);
+
+        if (queryInfo == null) {
+            return true;
+        }
+
+        DateTime queryEndTime = queryInfo.getQueryStats().getEndTime();
+
+        if (RUNNING.equals(queryInfo.getState()) || queryEndTime == null) {
```

Comment:
- `queryInfo.getState() == RUNNING ...`
  (i am using intellij's `Java | Code style issues | 'equals()' called on Enum value` inspection)
- please mark `com.facebook.presto.execution.QueryStats#getEndTime` be `@Nullable`